### DISCUSSION TITLE
How can you end a stage for multiple active players?

### DISCUSSION THREAD

**Discussion by zealvurte** [boardgameio/boardgame.io#947] [Q&A] [answered]
From what I can tell, it doesn't seem possible to programmatically end the stage for players other than the one taking an action (`ctx.playerID`).

The scenario I have is a vote, where everyone is put into the same stage to allow them to vote out of turn order. If anyone says no, I need to end that stage for everyone else, as the vote has already failed. Another use-case I was considering is where the first player to do something in a stage would win that stage and so the stage should end for everyone else too.

As this can happen in the middle of other stages, I'm using the revert option with `setActivePlayers` for this. Obviously the revert only kicks in when `activePlayers` becomes empty, and that behaviour isn't configurable (unlike something similar like turn order).

**Is there a known workaround to achieve something like this, or is it something new that needs support?**
I guess `endIf` for stages is the comparatively missing API here, but I'm sure there's other future solutions too.

**Do subsequent `setActivePlayers` with revert, allow a revert for each time it was called that way, or does it only keep track of one previous state and thus the revert only works for the most recent call?**
If the latter, I guess I should just build something to keep a history of the stages, and use setActivePlayers to manually revert back to the previous stages each time as a workaround, because the revert itself won't fulfil my needs anyway, but I want to check for alternatives before going down that path.

**Comment by delucis** [comment, marked as answer]
> Is there a known workaround to achieve something like this, or is it something new that needs support?

I think there may be some limitations with how stages work currently for this scenario. As you mentioned, `revert` (and also the `next` option) are only activated once `activePlayers` becomes empty, so to trigger them you need to call `endStage` for each player. Perhaps the missing API is something like an `endAllStages` event, to trigger that next/revert step early.

You can obviously end the stage for everyone using `setActivePlayers`, but this doesn’t play well with the `revert` option. You can possibly workaround this by using `next` instead of `revert`. The only limitation is that this won’t work correctly with move limits because we won’t be able to correctly track the number of moves per stage as revert does internally.

```js
// Enter the voting stage
setActivePlayers({
  all: 'vote',
  // Store the current active players as the next value
  next: {
    value: ctx.activePlayers,
    // `next` can also be nested if you need a revert-style stack
    // next: ctx._nextActivePlayers
  },
});

// End the voting stage, using the stored “next” value early
setActivePlayers(ctx._nextActivePlayers);
```

This could also be accomplished by storing next/revert values in `G`, which may be conceptually clearer but I guess the above version also progresses as expected if everyone does call `endStage`. `ctx._nextActivePlayers` is technically “internal” state, not strictly speaking intended for public use, but it should work pretty reliably in this case.

> Do subsequent `setActivePlayers` with revert, allow a revert for each time it was called that way?

Yes, each subsequent call stacks on top of the previous one. These are [stored in the `ctx._prevActivePlayers` array](https://github.com/boardgameio/boardgame.io/blob/49c2c12fd02fd3cb4fd5872c7593beb4b3954511/src/core/turn-order.ts#L48-L54). As soon as `setActivePlayers` is called without `revert`, that stack gets reset.
